File: src/js/core/expr/code/renderexp.js
'use strict'

//// Renderexp ////

var cwd = process.cwd() + '/';
var Interpreter = require(cwd + "core/semantics/code/interpreter.js");
var Enso = require(cwd + "enso.js");

var Renderexp;

var render = function(obj, args = Enso.EMap.new()) {
  var self = this, interp;
  interp = RenderExprC.new();
  return interp.dynamic_bind(function() {
    return interp.render(obj);
  }, args);
};

function RenderExpr(parent) {
  return class extends Enso.mix(parent, Interpreter.Dispatcher) {
    render(obj) {
      var self = this;
      return self.dispatch_obj("render", obj);
    };

    render_EBinOp(obj) {
      var self = this;
      return Enso.S(self.render(obj.e1()), " ", obj.op(), " ", self.render(obj.e2()));
    };

    render_EUnOp(obj) {
      var self = this;
      return Enso.S(obj.op(), " ", self.render(obj.e()));
    };

    render_EField(obj) {
      var self = this;
      return Enso.S(self.render(obj.e()), ".", obj.fname());
    };

    render_ESubscript(obj) {
      var self = this;
      return Enso.S(self.render(obj.e()), "[", self.render(obj.sub()), "]");
    };

    render_EVar(obj) {
      var self = this;
      return obj.name();
    };

    render_EConst(obj) {
      var self = this;
      return obj.val();
    };

    render_ENil() {
      var self = this;
      return "";
    }; }};

class RenderExprC extends Enso.mix(Enso.EnsoBaseClass, RenderExpr) {
  static new(...args) { return new RenderExprC(...args) };

};

Renderexp = {
  render: render,
  RenderExpr: RenderExpr,
  RenderExprC: RenderExprC,
};
module.exports = Renderexp ;

File: src/core/expr/code/renderexp.rb
require 'core/semantics/code/interpreter'

module Renderexp
  module RenderExpr
    
    include Interpreter::Dispatcher    
      
    def render(obj)
      dispatch_obj(:render, obj)
    end
  
    def render_EBinOp(obj)
      "#{render(obj.e1)} #{obj.op} #{render(obj.e2)}"
    end
  
    def render_EUnOp(obj)
      "#{obj.op} #{render(obj.e)}"
    end
  
    def render_EField(obj)
      "#{render(obj.e)}.#{obj.fname}"
    end
  
    def render_ESubscript(obj)
      "#{render(obj.e)}[#{render(obj.sub)}]"
    end
  
    def render_EVar(obj)
      "#{obj.name}"
    end
  
    def render_EConst(obj)
      "#{obj.val}"
    end
    
    def render_ENil
      ""
    end
  end
  
  class RenderExprC
    include RenderExpr
  end

  def self.render(obj, args={})
    interp = RenderExprC.new
    interp.dynamic_bind(args) do
      interp.render(obj)
    end
  end
end
